libnetwork/drivers/bridge/port_mapping_linux.go

```diff
@@ -868,6 +875,59 @@ func setPerPortForwarding(b portBinding, ipv iptables.IPVersion, bridgeName stri
 	return nil
 }
 
+// filterPortByInputIface adds a couple of iptables rules to accept packets
+// destined to a mapped port with a dest addr that matches the interface
+// they're received on. If that doesn't match, the second rule drop them
+// unceremoniously.
+//
+// These rules will block rogue hosts that try to access a mapped port while
+// they aren't part of the L2 segment where the mapped port is exposed.
+// For instance, if HostIP == 127.0.0.1, no ingress should come from anything
+// but lo.
+func (n *bridgeNetwork) filterPortByInputIface(b portBinding, enable bool) error {
+	// DOCKER_DISABLE_INPUT_IFACE_FILTERING can be used as an escape hatch if
+	// this filtering doesn't work out well for some users.
+	if v, _ := strconv.ParseBool(os.Getenv("DOCKER_DISABLE_INPUT_IFACE_FILTERING")); v {
+		return nil
+	}
```

Comment:
> - Do we consider this one to be permanent or temporary ?

That's a temporary escape hatch in case the new filtering does break someone, somewhere.

> Should we have a log entry to indicate this option was set?

I added an `info` log to that code path. Now, whenever users create a container with NAT-based port-mappings, they'll get this logged:

```
DOCKER_DISABLE_INPUT_IFACE_FILTERING is set, skipping input iface filtering
```

> If we consider it temporary, have a warning added to docker info ? (check if it's set, and if so, add a warning to the list of warnings returned)?

I added these two warnings to make it clear that it's insecure, and that it'll be removed at a later time.

```
WARNING: input interface filtering is disabled on port mappings, this might be insecure
DEPRECATED: DOCKER_DISABLE_INPUT_IFACE_FILTERING is deprecated and will be removed in a future release
```